spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/ResultSetWrapper.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.jdbc.core.convert;
+
+import lombok.val;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.mapping.MappingException;
+import org.springframework.lang.Nullable;
+
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Wraps a {@link java.sql.ResultSet} in order to provide fast lookup of columns by name, including for missing columns.
+ *
+ * @author Jens Schauder
+ */
+class ResultSetWrapper {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ResultSetWrapper.class);
+
+	private final ResultSet resultSet;
+
+	private final Map<String, Integer> indexLookUp;
+
+	ResultSetWrapper(ResultSet resultSet) {
+
+		this.resultSet = resultSet;
+		indexLookUp = indexColumns();
+	}
+
+	private Map<String, Integer> indexColumns() {
+
+		try {
+
+			ResultSetMetaData metaData = resultSet.getMetaData();
+			int columnCount = metaData.getColumnCount();
+
+			Map<String, Integer> index = new HashMap<>(columnCount);
+
+			for (int i = 1; i <= columnCount; i++) {
+
+				String label = metaData.getColumnLabel(i).toLowerCase();
+				Integer previous = index.put(label, i);
```

Comment:
As per JDBC spec: Encountering a column label more than once, then the first label is used. I'd suggest a case-insensitive `Set<String>` (see `LinkedCaseInsensitiveMap`) to check for column names.